src/Product/ProductFilter.php

```diff
@@ -67,4 +67,28 @@ public function filter_sync_ready_products( array $products, bool $return_ids =
 
 		return new FilteredProductList( $results, $unfiltered_count );
 	}
+
+	/**
+	 * Filter and return a list of products that can be deleted.
+	 *
+	 * @since x.x.x
+	 *
+	 * @param WC_Product[] $products
+	 * @param boolean      $return_ids
+	 *
+	 * @return array
+	 */
+	public function filter_products_for_delete( array $products, bool $return_ids = false ): array {
```

Comment:
💅

Not sure `$return_ids` is needed here. Method is only used at one single place and returns ids. Filters also filter products, so, maybe, should always return products and then, if you need ids, map to get ids outside of filter. This is still going to be another loop, but same time it is not going to raise the order of complexity.

We could also pass `Helper` as a filtering condition parameter into filter methods instead of class dependency and remove `$return_ids` completely. Besides there seems to be a bug in this class with `filter_sync_ready_products` using `$return_ids` when `woocommerce_gla_get_sync_ready_products_filter`
```
/**
	* Filters the list of products ready to be synced (after applying filters to check failures and sync-ready status).
	*
	* @param WC_Product[] $results Sync-ready WooCommerce products
	*/
$results = apply_filters( 'woocommerce_gla_get_sync_ready_products_filter', $results );
```
expects Product objects, not ids.

Reply:
A small typo here, which is not a part of this PR, or course, but I found it.

https://github.com/woocommerce/google-listings-and-ads/blob/3b2dd688aeaff4cbddce20112e6c159d6e5b3eb4/src/Product/ProductFilter.php#L25

Reply:
Nice catch, yes it should always be a set of products for that filter. We aren't actually using the filter `woocommerce_gla_get_sync_ready_products_filter` within our plugin (only the pre filter), which is probably why it went unnoticed. I'd suggest to report these two problems in a separate issue, it would be nice to take a deeper look at the unit tests, to see if this could be caught there as well.

Reply:
Created this one https://github.com/woocommerce/google-listings-and-ads/issues/1259